File: chess-game/backend/chessapp/domain/pieces/king.py
import math
from chessapp.domain.rules.movements.king_rule import KingRule
from chessapp.domain.pieces.piece import Piece
from chessapp.domain.pieces.piece_type import PieceType
from chessapp.domain.value_objects.piece_id import PieceId
from chessapp.domain.value_objects.side import Side

class King(Piece):

    def get_points(self):
        return math.inf

    def get_rule(self):
        return KingRule()

    def __init__(self, piece_id: PieceId, side: Side):
        super().__init__(piece_id, side, PieceType.King)
        self._has_moved = False
        self._has_castled = False

    def move(self):
        self._has_moved = True

    def has_moved(self):
        return self._has_moved

    def castle(self):
        self._has_castled = True

    def has_castled(self):
        self._has_castled = True

File: frontend/src/app/pages/game/play/play/models/pieces/king.ts
import { Cell } from "../board/ cell";
import { Side } from "../side";
import { Piece } from "./piece";
import { PieceType } from "./piece_type";

export class King extends Piece {
    
    override get can_move_over(): boolean {
        return false;
    }
    
    constructor(side: Side) {
        super(side, PieceType.King);
    }

    override validateMove(from: Cell, to: Cell): boolean {
        const [delta_file, delta_rank] = this.calculateMoveDeltas(from, to);

        return Math.abs(delta_file) <= 1 && Math.abs(delta_rank) <= 1;
    }
}